File: OOP/askisi3/plane.cpp
/* File: plane.cpp */

#include <iostream>
#include <string>

#include "plane.h"

using namespace std;

Plane::Plane(string title, int capacity) : title(title) , capacity(capacity) {
	this->carBay = new CargoBay;
	for ( int i = 0 ; i < num_of_equipComp ; i++ )
		this->equipComp[i] = new EquipmentCompartment;
	this->passComp = new PassengerCompartment;
	cout << "Plane just created!" << endl;
}

Plane::~Plane() {	// delete all dynamic allocated memory
	for ( int i = 0 ; i < num_of_equipComp ; i++ )
		delete this->equipComp[i];
	delete this->passComp;
	delete this->carBay;
	cout << "Plane to be destroyed!" << endl;
}

String *Plane::get_title() {
	return &this->title;
}

int Plane::get_capacity() const {
	return this->capacity;
}

CargoBay *Plane::get_carBay() {
	return this->carBay;
}

EquipmentCompartment **Plane::get_equipComp() {
	return this->equipComp;
}

PassengerCompartment *Plane::get_passComp() {
	return this->passComp;
}

void Plane::process(SecurityEmployee* securEmployee) {		// plane's process calls process of each PlaneComponent depending on the Employee
	this->passComp->process(securEmployee);
	for ( int i = 0 ; i < num_of_equipComp ; i++ )
		this->equipComp[i]->process(securEmployee);	// work to all EquipmentCompartments
	this->carBay->process(securEmployee);
}

void Plane::process(CleaningEmployee* cleanEmployee) {
	this->passComp->process(cleanEmployee);
	this->carBay->process(cleanEmployee);
}

void Plane::process(MaintenanceEmployee* mainEmployee) {
	for ( int i = 0 ; i < num_of_equipComp ; i++ )
		this->equipComp[i]->process(mainEmployee);
	this->carBay->process(mainEmployee);
}

void Plane::ready_check() {					// if all PlaneComponents are ready then plane is ready
	for ( int i = 0 ; i < num_of_equipComp ; i++ )		// first chech the EquipmentCompartments
		if ( this->equipComp[i]->ready_check() == false ) {
			cout << "Plane not ready to take off!" << endl;
			return;
		}						// then the rest
	if ( this->passComp->ready_check() == false || this->carBay->ready_check() == false ) {
			cout << "Plane not ready to take off!" << endl;
			return;
		}
	cout << "Plane ready to take off!" << endl;
}

Plane *Plane::clone() {
	Plane *newplane = new Plane(this->title.get_string(), this->capacity);
	newplane->id = this->id;
	newplane->title = *this->title.clone();
	newplane->carBay = this->carBay->clone();
	newplane->passComp = this->passComp->clone();
	for ( int i = 0 ; i < num_of_equipComp ; i++ )
		newplane->equipComp[i] = this->equipComp[i]->clone();
	return newplane;
}

int Plane::equal(Object *plane) {				// casting Object * to Plane * so we can access Plane's data-members
	if ( dynamic_cast<Plane *>(plane) == NULL )		// if cast fails then return 0(not equal)
		return 0;
	if ( this->id == dynamic_cast<Plane *>(plane)->id )
		if ( this->title.equal(&dynamic_cast<Plane *>(plane)->title) == 1 )
			if ( this->carBay->equal(dynamic_cast<Plane *>(plane)->carBay) == 1 )
				if ( this->passComp->equal(dynamic_cast<Plane *>(plane)->passComp) == 1 ) {
					for ( int i = 0 ; i < num_of_equipComp ; i++ )
						if ( dynamic_cast<Plane *>(plane)->equipComp[i] == NULL )
							return 0;
						else if ( this->equipComp[i]->equal(dynamic_cast<Plane *>(plane)->equipComp[i]) == 0 )
							return 0;
					return 1;
				}
	return 0;
}

String Plane::toString() {
	String str = String(this->equipComp[0]->toString().get_string());
	String comma = String(", ");
	for ( int i = 1; i < num_of_equipComp ; i++ ) {		// first create a String which contains each EquipmentCompartment's toString
		String *tempstr = new String(this->equipComp[i]->toString());
		str.concat(comma);
		str.concat(*tempstr);
		delete tempstr;
	}							// Plane's toString returns a String which contains each data-member's toString
	return String(Object::toString().get_string() + ", title = " + title.toString().get_string() + ", capacity = " + std::to_string(this->capacity) + ", carBay = " + this->carBay->toString().get_string() + ", passComp = " + this->passComp->toString().get_string() + ", equipComp[0-" + std::to_string(num_of_equipComp-1) + "] = " + str.get_string());
}

File: OOP/askisi4/Plane.java
// File: Plane.java

package askisi4;

public class Plane {

    private String title;
    private int capacity;
    private CargoBay carBay;
    private EquipmentCompartment equipComp[];
    private PassengerCompartment passComp;

    public Plane(String title, int capacity) {
        this.title = title;
        this.capacity = capacity;
        this.carBay = new CargoBay();
        this.equipComp =  new EquipmentCompartment[3];
        for ( int i = 0 ; i < 3 ; i++ )
            this.equipComp[i] = new EquipmentCompartment();
        this.passComp = new PassengerCompartment();
    }

    public String get_title() {
        return this.title;
    }

    public int get_capacity() {
        return this.capacity;
    }

    public CargoBay get_carBay() {
        return this.carBay;
    }

    public PassengerCompartment get_passComp() {
        return this.passComp;
    }

    public EquipmentCompartment[] get_equipComp() {
        return this.equipComp;
    }

    // plane's process calls process of each PlaneComponent depending on the Employee

    public void process(SecurityEmployee sEmp) {
        this.passComp.process(sEmp);
        for ( int i = 0 ; i < 3 ; i++ )
            this.equipComp[i].process(sEmp);
        this.carBay.process(sEmp);
    }

    public void process(CleaningEmployee cEmp) {
        this.passComp.process(cEmp);
        this.carBay.process(cEmp);
    }

    public void process(MaintenanceEmployee mEmp) {
        for ( int i = 0 ; i < 3 ; i++ )
            this.equipComp[i].process(mEmp);
        this.carBay.process(mEmp);
    }

    public void ready_check() {     		// if all PlaneComponents are ready then plane is ready
        for ( int i = 0 ; i < 3 ; i++ )
            if ( this.equipComp[i].ready_check() == false ) {
                System.out.println("Plane not ready to take off!");
                return;
            }
        if ( this.passComp.ready_check() == false || this.carBay.ready_check() == false ) {
            System.out.println("Plane not ready to take off!");
            return;
        }
        System.out.println("Plane ready to take off!");
    }

    @Override
    public String toString() {  		// Plane's toString returns a String which contains each data-member's toString
        return "Plane: title = " + this.title
                + ", capacity = " + this.capacity
                + ", carBay = " + this.carBay.toString()
                + ", equipComp[0-3] = " + this.equipComp[0].toString() + ", "
                + this.equipComp[1].toString() + ", " + this.equipComp[2].toString()
                + ", passComp = " + this.passComp.toString();
    }

    @Override
    public boolean equals(Object obj) {
        if ( !(obj instanceof Plane) )          // check if obj is of Plane type
            return false;                       // if not return, else check the equality
        for ( int i = 0 ; i < 3 ; i++ )         // casting Object to Plane so we can access its data-members
            if ( this.equipComp[i].equals(((Plane) obj).equipComp[i]) == false )
                return false;
        if ( this.capacity == ((Plane) obj).capacity )
            if ( this.title.equals(((Plane) obj).title) == true )
                if ( this.carBay.equals(((Plane) obj).carBay) == true )
                    if ( this.passComp.equals(((Plane) obj).passComp) == true )
                        return true;
        return false;
    }

    @Override
    public Object clone() { 			// casting because clone returns Object
        Plane nPlane = new Plane(this.title, this.capacity);
        nPlane.carBay = (CargoBay) this.carBay.clone();
        nPlane.passComp = (PassengerCompartment) this.passComp.clone();
        for ( int i = 0 ; i < 3 ; i ++ )
            nPlane.equipComp[i] = (EquipmentCompartment) this.equipComp[i].clone();
        return nPlane;
    }
}
